Implement signal prototype

AnOctopus opened the pull request.

AnOctopus committed:
Implement signal prototype

Widgets have additional arguments for setting a signal, with an optional
context value, when they change, and an API for detecting signals and getting
the updated widget and context values is provided.

kmcgrady commented on lib/streamlit/widgets.py:
```diff
@@ -104,26 +110,57 @@ def call_callbacks(self, new_widget_states: WidgetStates) -> None:
                 continue
 
             # The widget's value has changed - call its on_change callback.
             callback(new_value)
```
`new_value` is not defined now.

kmcgrady replied:
Trying this out in PR Preview https://share.streamlit.io/streamlit/core-previews/pr-2884

```
import streamlit as st
state = st.beta_session_state(count=0)

if st.signal('Increment'):
    increment_value = st_event.context()
    state.count += increment_value
  

selected_value = st.selectbox('Select me', ['cat', 'dog'])
st.button('Button', signal='Increment', context=1)

st.write('Counter Value = ', state.count)
```

AnOctopus replied:
Looks like my editor broke, should have highlighted that `new_value` was no longer defined

kmcgrady reviewed (changes requested):
Fixing the error that is popping up is necessary. Using the example code in the thing. I think there's one missing imlementation.

> A rerun, not triggered by a widget or triggered by a widget with no signal should reset signal to return false.

Essentially. If I click "Increment", it increments correctly. If I change the selectbox, the Increment signal should no longer be set. Otherwise, changing the select box will keep incrementing.

AnOctopus committed:
Remove debugging print statements

AnOctopus committed:
Fix undefined variable in widget callback code

AnOctopus committed:
Reset signals from previous script runs

The assumption that it was already cleared was conflating the widget signal
arguments and the global signal state.
